osu.Game.Rulesets.Osu/Mods/OsuModBubbles.cs

```diff
@@ -0,0 +1,237 @@
+            ((PlaySliderBody)slider.Body.Drawable).BorderColour = slider.AccentColour.Value;
+
+        #region Pooled Bubble drawable
+
+        private partial class BubbleDrawable : PoolableDrawable
+        {
+            private readonly Box colourBox;
+            private readonly CircularContainer content;
+
+            public BubbleInfo Info { get; set; }
+
+            public BubbleDrawable()
+            {
+                Origin = Anchor.Centre;
+                InternalChild = content = new CircularContainer
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    RelativeSizeAxes = Axes.Both,
+                    MaskingSmoothness = 2,
+                    BorderThickness = 0,
+                    BorderColour = Colour4.White,
+                    Masking = true,
+                    EdgeEffect = new EdgeEffectParameters
+                    {
+                        Type = EdgeEffectType.Shadow,
+                        Radius = 3,
+                        Colour = Colour4.Black.Opacity(0.05f),
+                    },
+                    Child = colourBox = new Box { RelativeSizeAxes = Axes.Both, }
+                };
+            }
+
+            protected override void PrepareForUse()
+            {
+                Alpha = 1;
+                Colour = Colour4.White;
+                Scale = new Vector2(1);
+                Position = Info.Position;
+                Size = Info.InitialSize;
+                content.BorderThickness = Info.InitialSize.X / 3.5f;
+                content.BorderColour = Colour4.White;
+
+                //We want to fade to a darker colour to avoid colours such as white hiding the "ripple" effect.
+                ColourInfo colourDarker = Info.Colour.Darken(0.1f);
+
+                // Main bubble scaling based on combo
+                this.ScaleTo(Info.MaxSize, getAnimationDuration() * 0.8f)
+                    .Then()
+                    // Pop at the end of the bubbles life time
+                    .ScaleTo(Info.MaxSize * 1.5f, getAnimationDuration() * 0.2f, Easing.OutQuint)
+                    .FadeOutFromOne(getAnimationDuration() * 0.2f, Easing.OutCirc).Expire();
+
+                if (Info.IsHit)
+                {
+                    colourBox.FadeColour(colourDarker);
+
+                    content.TransformTo(nameof(BorderColour), colourDarker, getAnimationDuration() * 0.3f, Easing.OutQuint);
+                    // Ripple effect utilises the border to reduce drawable count
+                    content.TransformTo(nameof(BorderThickness), 2f, getAnimationDuration() * 0.3f, Easing.OutQuint)
+                           // Avoids transparency overlap issues during the bubble "pop"
+                           .Then().Schedule(() => content.BorderThickness = 0);
+
+                    return;
+                }
+
+                Colour = Colour4.Black;
```

Comment:
Or black?

Reply:
Black is specifically for missed hits. I don't get what you're trying to point out ?

Reply:
The structure must be very weird then, because both the colour specifications are at the same level in the same method. I'd recommend restructuring this because I could not notice this.